feat [#446] 검색 API 통합

ch1hyun opened the pull request:
<!-- 제목양식을 지켜주세요! Prefix [#이슈번호] {PR 설명} -->
<!-- PR 작성 후 우측에 Development에서 이슈 찾아서 연동하면 merge될때 이슈도 close됩니다 -->
<!-- Reviewer, Assignees, Label 붙이기 --> 
## 🔗 Issue Number
<!-- #{본인 이슈 번호} 치면 알아서 이슈 게시판 링크 걸려요 -->
closes #446 

## 🔙 Previous PR
<!-- 이어지는 PR이라면 이전 PR 링크를 남겨주세요 (ex: #123) -->
- #337 
- #345 
- #341 

## 📌  To-do
<!-- 본인이 한 업무를 체크리스트로 작성해주세요 -->

- [x] 통합 검색 API 구현
- [x] 이전 검색 API 제거


## ✨Description 
<!-- 본인이 한 작업을 설명해주세요 -->
아래의 검색 방법 3가지를 사용해 결과가 하나의 API에서 조회될 수 있도록 변경했어요.
1. 아티스트 연관 검색
2. 공연 연관 검색
3. 검색어 검색

[ 아티스트 연관 검색 ]
<img width="1074" alt="image" src="https://github.com/user-attachments/assets/15a279c2-92fb-4ba0-8366-7a115adfa9e7" />

[ 공연 연관 검색 ]
<img width="1074" alt="image" src="https://github.com/user-attachments/assets/13ccd638-ab9b-4f1a-a05b-b44d56717374" />

[ 검색어 검색 ]
<img width="1074" alt="image" src="https://github.com/user-attachments/assets/359a760c-bde3-4566-abdf-45597c0d97ea" />

ch1hyun committed:
refactor [#446] 클래스 명 변경

- SearchTermController -> SearchController

ch1hyun committed:
refactor [#446] 클래스 명 변경

- SearchTermController -> SearchController

ch1hyun committed:
feat [#446] 홈 검색 API 아티스트 아이디로 조회하는 기능 구현

ch1hyun committed:
feat [#446] 홈 검색 API 아티스트 아이디로 조회하는 기능 구현

ch1hyun committed:
feat [#446] 홈 검색 API 아티스트 아이디로 조회하는 기능에서 검색어 저장

ch1hyun committed:
feat [#446] 홈 검색 API 아티스트 아이디로 조회하는 기능에서 검색어 저장

ch1hyun committed:
feat [#446] 홈 검색 API 공연 아이디로 조회하는 기능 구현

ch1hyun committed:
feat [#446] 홈 검색 API 공연 아이디로 조회하는 기능 구현

ch1hyun committed:
feat [#446] 검색어 분석 유틸 클래스 작성

ch1hyun committed:
feat [#446] 검색어 분석 유틸 클래스 작성

ch1hyun committed:
feat [#446] 홈 검색 API 검색어로 조회하는 기능 구현

ch1hyun committed:
feat [#446] 홈 검색 API 검색어로 조회하는 기능 구현

ch1hyun committed:
refactor [#446] 더 이상 사용되지 않는 검색 API 제거

ch1hyun committed:
refactor [#446] 더 이상 사용되지 않는 검색 API 제거

Ivoryeee commented on src/main/java/org/sopt/confeti/api/search/facade/SearchFacade.java:
```diff
@@ -0,0 +1,151 @@
+                        p -> false
+                ));
+
+        if (Objects.nonNull(userId)) {
+            List<PerformanceDTO> favoritePerformances = performanceService.getFavoritePerformancesAll(userId, TYPE_ALL);
+            getPerformanceFavorites(performanceFavorites, favoritePerformances);
+        }
```
`if (Objects.nonNull(userId))` 블록이 두 번 반복되는 것 같은데, 의도한 바가 있으실까요?
아니라면 userId가 null일 경우의 로직으로 정리하거나, 아래와 같이 userId가 존재하는 경우를 하나의 if문에서 묶어서 작성하는 방식은 어떠신가요?

```java
    if (Objects.nonNull(userId)) {
        artistFavorite = artistFavoriteService.isFavorite(userId, aid);

        List<PerformanceDTO> favoritePerformances = performanceService.getFavoritePerformancesAll(userId, TYPE_ALL);
        getPerformanceFavorites(performanceFavorites, favoritePerformances);
    }
```

ch1hyun replied:
아티스트를 조회하는 블록, 공연 목록을 조회하는 블록을 나누었었는데요, 맥락을 명확히 표현하고자 했습니다!

아티스트를 조회할 때 아티스트 좋아요 계산
공연을 조회할 때 공연 좋아요 계산

응집성있는 코드를 작성하는 것이 클린 코드라고 생각했어요

Ivoryeee commented on src/main/java/org/sopt/confeti/api/search/facade/SearchFacade.java:
```diff
@@ -0,0 +1,151 @@
+        searchTermService.write(artist.getName());
+        boolean artistFavorite = false;
+
+        if (Objects.nonNull(userId)) {
```
단순 궁금증인데 `userId != null` 대신 `Objects.nonNull(userId)`로 작성하신 의도가 있는지 궁금해요! 
전자가 직관적이라 생각해 `userId != null`로 작성해 왔는데 `Objects.nonNull`로 작성하셨길래 한번 여쭙고 싶습니다 ㅎ.ㅎ

ch1hyun replied:
!= 는 null을 직접 작성해야하고 논리적 문법을 사용해 계산돼요. 그리고 객체지향적 관점으로 null 여부를 검사하는 책임을 가지게 되기 때문에 이를 해결할 수 있는 Objects 클래스에 null 검사 메세지를 요청해 처리하도록 했습니다. (여기서는 nonNull)

그리고 메서드 이름으로 어떤 조건인지 나타내면 의도가 명확해진다고 생각했어요.

Ivoryeee commented on src/main/java/org/sopt/confeti/api/search/facade/SearchFacade.java:
```diff
@@ -0,0 +1,151 @@
+    }
+
+    @Transactional(readOnly = true)
+    public SearchResultDTO getHomeSearchResultWithTerm(Long userId, String term) {
```
요 부분 메서드 내부 길이가 조금 길어졌는데, `Objects.nonNull(userId)`조건을 하나로 묶거나 별도의 메서드로 분리하면 좋을 것 같아요!

ch1hyun replied:
트랜잭션 적용을 위해 따로 분리하지 않았었어요

혹시 좋은 아이디어가 있으시다면 개선 코드를 작성해주시면 참고해서 수정할게요!

Ivoryeee reviewed (approved):
역시 코드 분리가 깔끔하시네요... 👍🏻 👍🏻  너무 수고 많으셨습니다!

junggyo1020 commented on src/main/java/org/sopt/confeti/api/search/controller/SearchController.java:
```diff
@@ -0,0 +1,75 @@
+
+        SearchResultDTO searchResult = null;
+
+        if (Objects.nonNull(aid)) {
+            searchResult = searchFacade.getHomeSearchResultWithAid(userId, aid);
+        }
+
+        if (Objects.isNull(searchResult) && Objects.nonNull(pid)) {
+            searchResult = searchFacade.getHomeSearchResultWithPid(userId, pid);
+        }
+
+        if (Objects.isNull(searchResult) && Objects.nonNull(term)) {
+            searchResult = searchFacade.getHomeSearchResultWithTerm(userId, term);
+        }
```
현재 Controller 단에서 조건 분기를 if (Objects.nonNull()) 과 같이 3단계로 나누고 있는데, 
이 조건 분기를 Enum 기반으로 전략 패턴을 적용하거나 라우터 메서드로 추출하면 테스트가 훨씬 수월해질 것 같아요.
```java
private enum SearchType {
  AID, PID, TERM
}
```
내부 헬퍼 메서드로 SearchType resolveType()으로 분리하면 책임이 명확해질 것 같아요. 
지금처럼 if 를 이용한 체이닝 보다 가독성 및 유지보수에 용이할 것 같아서 제시해보았습니다 :)

ch1hyun replied:
Enum 기반으로 전략 패턴을 활용하는 방법이 잘 생각이 안나서 그런데.. 혹시 코드 제시 해주실 수 있나요..?
정말 좋은 방법인 것 같은데 로직이 그려지지 않네요.. ㅜ 🥲

혹시 아래와 같은 코드를 말씀하셨던 걸까요?
```JAVA
SearchType searchType = SearchType.resolveType(aid, pid, term);
if (searchType.equals(SearchType.AID)) ~~
else if (searchType.equals(SearchType.PID)) ~~
else if (searchType.equals(SearchType.TERM)) ~~
```

junggyo1020 replied:
전략패턴을 사용하면 아래처럼 사용할 수 있을 것 같아요.
```java
public enum SearchType {
    AID {
        @Override
        public SearchResultDTO search(SearchFacade facade, Long userId, String aid, String pid, String term) {
            return facade.getHomeSearchResultWithAid(userId, aid);
        }
    },
    PID {
        @Override
        public SearchResultDTO search(SearchFacade facade, Long userId, String aid, String pid, String term) {
            return facade.getHomeSearchResultWithPid(userId, pid);
        }
    },
    TERM {
        @Override
        public SearchResultDTO search(SearchFacade facade, Long userId, String aid, String pid, String term) {
            return facade.getHomeSearchResultWithTerm(userId, term);
        }
    },
    NONE {
        @Override
        public SearchResultDTO search(SearchFacade facade, Long userId, String aid, String pid, String term) {
            return null;
        }
    };

    public abstract SearchResultDTO search(SearchFacade facade, Long userId, String aid, String pid, String term);

    public static SearchType resolve(String aid, String pid, String term) {
        if (Objects.nonNull(aid)) return AID;
        if (Objects.nonNull(pid)) return PID;
        if (Objects.nonNull(term)) return TERM;
        return NONE;
    }
}
```
이렇게 전략패턴으로 작성하면 컨트롤러 단에서는 아래처럼 호출만하면 될 것 같아요
```java
SearchType searchType = SearchType.resolve(aid, pid, term);
SearchResultDTO searchResult = searchType.search(searchFacade, userId, aid, pid, term);
```
이렇게 작성했을 때 장점은 조건문 없이 작성하면서도 명확하게 책임을 분리할 수 있고, 새로운 검색 유형을 추가할 일이 있다면 Enum만 수정할 수 있어 확장에 열려있으면서, search() 메서드에 전략 위임해서 SRP 원칙을 더 잘 지킬 수 있다고 생각했어요!

제 코드가 정답은 아니니, 참고만 해주세요 :)

ch1hyun replied:
오호 정말 좋은 의견이라고 생각합니다!
말씀하신대로 명확한 책임 분리와 확장성을 가질 수 있네요. 의견 반영해서 리펙토링 할게요!

junggyo1020 reviewed (approved):
수고하셨습니다!
전체 Search 관련 기능을 통합한 부분이 좋은 리펙토링 방향인 것 같아요 ㅎㅎ

Ivoryeee committed:
refactor [#452] profileUrl(String) -> profileFile(MultipartFile) 변경 (#453)

ch1hyun committed:
refactor [#446] 검색 쿼리에 따른 분기 처리 리펙토링

ch1hyun committed:
refactor [#446] 검색 쿼리에 따른 분기 처리 리펙토링

ch1hyun committed:
setting [#436] Spring Rest Docs + Open API 명세 생성 및 Swagger 연동

* feat [#436] 테스트 기본 데이터 작성

* feat [#436] 엘라스틱 서치 인덱스, 매핑 구성 파일 작성

* feat [#436] Test Container 사용을 위한 세팅

* feat [#436] 테스트용 Elastic Search 구성 정보 작성

* feat [#436] 테스트 기본 추상 클래스 작성

* feat [#436] Test Container, Spring Rest Docs 및 Open API 자동 생성을 위한 라이브러리 추가

* feat [#436] 인덱스 구성 정보 파일 연결

* refactor [#436] 기본 값 수정

* fix [#436] 검색어로 아티스트를 검색했을 때도 기록에 수집되도록 추가

* feat [#436] 전역 테스트 컨테이너 생성 및 관리로 수정

* refactor [#436] 하이버네이트 캐시 오류로 매번 재생성되도록 수정

* feat [#436] 각 테스트마다 환경의 멱등성을 보장하기 위해 데이터 초기화 로직 추가

* feat [#436] 예제 테스트 작성

- Artist Search
- User Favorite
- User Timetable

* feat [#436] Redis Test Container 추가

* refactor [#436] 로깅 삭제

* refactor [#436] 폴더 경로 수정 및 역할에 맞게 클래스 분리

* feat [#436] Swagger UI 연동

* feat [#436] Swagger UI 의존성 추가

* refactor [#436] openapi3.yml 노출 제거

* style [#436] 공백 추가

* build [#436] CI/CD Work Flow에 application-test.yml 구성 추가

* RestDocs 적용으로 사용되지 않는 테스트 코드 주석처리

이후 테스트 코드 작성 시 참고할 예정이라 삭제가 아닌 주석처리 하였습니다.

* chore [#436] 추후 사용해야 할 테스트 코드 주석처리

* chore [#436] 추후 사용해야 할 테스트 코드 주석처리

* chore [#436] 사용하지 않는 테스트 코드 삭제

* build [#436] CI/CD 워크플로우에 테스트 컨테이너 reuse 설정 추가

* build [#436] CI/CD 워크플로우에 오류 리포트 확인 추가

* build [#436] CI/CD 워크플로우 빌드 디버깅 추가

* build [#436] CI/CD 워크플로우 도커 허용

* build [#436] CI/CD 워크플로우 수정

* refactor [#436] 사용되지 않는 테스트 제거

* refactor [#436] CI/CD 필요없는 구성 정보 제거

* refactor [#436] CD Test Container reuse 로컬 우선순위 강화

* refactor [#452] profileUrl(String) -> profileFile(MultipartFile) 변경 (#453)

* refactor [#436] 유저 디스코드 알림 연동 감시를 위한 로그 추가

* feat [#436] 테스트 기본 데이터 작성

* feat [#436] 엘라스틱 서치 인덱스, 매핑 구성 파일 작성

* feat [#436] Test Container 사용을 위한 세팅

* feat [#436] 테스트용 Elastic Search 구성 정보 작성

* feat [#436] 테스트 기본 추상 클래스 작성

* feat [#436] Test Container, Spring Rest Docs 및 Open API 자동 생성을 위한 라이브러리 추가

* feat [#436] 인덱스 구성 정보 파일 연결

* refactor [#436] 기본 값 수정

* fix [#436] 검색어로 아티스트를 검색했을 때도 기록에 수집되도록 추가

* feat [#436] 전역 테스트 컨테이너 생성 및 관리로 수정

* refactor [#436] 하이버네이트 캐시 오류로 매번 재생성되도록 수정

* feat [#436] 각 테스트마다 환경의 멱등성을 보장하기 위해 데이터 초기화 로직 추가

* feat [#436] 예제 테스트 작성

- Artist Search
- User Favorite
- User Timetable

* feat [#436] Redis Test Container 추가

* refactor [#436] 로깅 삭제

* refactor [#436] 폴더 경로 수정 및 역할에 맞게 클래스 분리

* feat [#436] Swagger UI 연동

* feat [#436] Swagger UI 의존성 추가

* refactor [#436] openapi3.yml 노출 제거

* style [#436] 공백 추가

* build [#436] CI/CD Work Flow에 application-test.yml 구성 추가

* RestDocs 적용으로 사용되지 않는 테스트 코드 주석처리

이후 테스트 코드 작성 시 참고할 예정이라 삭제가 아닌 주석처리 하였습니다.

* chore [#436] 추후 사용해야 할 테스트 코드 주석처리

* build [#436] CI/CD 워크플로우에 테스트 컨테이너 reuse 설정 추가

* build [#436] CI/CD 워크플로우에 오류 리포트 확인 추가

* build [#436] CI/CD 워크플로우 빌드 디버깅 추가

* build [#436] CI/CD 워크플로우 도커 허용

* build [#436] CI/CD 워크플로우 수정

* refactor [#436] 사용되지 않는 테스트 제거

* refactor [#436] CI/CD 필요없는 구성 정보 제거

* refactor [#436] CD Test Container reuse 로컬 우선순위 강화

* refactor [#436] 유저 디스코드 알림 연동 감시를 위한 로그 추가

---------

Co-authored-by: Willy <willy_1020@naver.com>
Co-authored-by: Ivoryeee <105477246+Ivoryeee@users.noreply.github.com>

ch1hyun committed:
Merge remote-tracking branch 'origin/feat#446/integrateSearchAPI' into feat#446/integrateSearchAPI

ch1hyun committed:
refactor [#446] CI/CD Workflow에 Swagger 배포를 위한 폴더 생성 추가

ch1hyun committed:
refactor [#446] 사용되지 않는 API 테스트 제거
